### PR TITLE
migrate to gojmx

### PR DESCRIPTION
**Pull request by cristianciutea**

### CODE REVIEW

File: src/connection/jmx.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		err = fmt.Errorf("failed to open new connection, %w", err)
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			p.sem.Release(1)
+		}
+	}()
```
Comment (roobre) [inline src/connection/jmx.go:65]: I think I get what you want to achieve here (release after this func is called if there's an error) but I think doing it this way is not very intuitive for the reader.

Is there a different way in which we could do this? e.g. locking the semaphore after we check that `gojmx.NewClient` returns successfully?

**Review by roobre — CHANGES_REQUESTED** [review]
Left a few nits here and there. However, the untested singleton dispatcher a blocker for me in order to merge this. I'd love if we could rewrite it into something less global and have a small test for it.

**Review by roobre — APPROVED** [review]
Unblocking this now, but I'd love if someone else on @newrelic/coreint could give it some 👀 as well

File: src/client/producer_consumer_collection.go [hunk 2]
```diff
@@ -64,23 +64,26 @@ func ConsumerWorker(consumerChan <-chan *args.JMXHost, wg *sync.WaitGroup, i *in
...
+			}
 
-			if err := jmxwrapper.JMXOpen(jmxInfo.Host, strconv.Itoa(jmxInfo.Port), jmxInfo.User, jmxInfo.Password, options...); err != nil {
+			conn, err := jmxConnProvider.NewConnection(config)
+			if err != nil {
 				log.Error("Unable to make JMX connection for Consumer '%s': %s", consumerEntity.Metadata.Name, err.Error())
-				jmxwrapper.JMXClose() // Close needs to be called even on a failed open to clear out any set variables
-				jmxwrapper.JMXLock.Unlock()
 				continue
 			}
```
Comment (alvarocabanas) [inline src/client/producer_consumer_collection.go:88]: I would also move this and in the producerWorker bellow Jmx bellow to a private getConsumerJMXConnection or ConsumerJMXConnect function, just to have it more readable.

**Review by alvarocabanas — APPROVED** [review]
Nice work! I just put a couple of nits just to separate a bit the concerns, but completely optional.

**Comment by cristianciutea** [comment]
> Left a few nits here and there. However, the untested singleton dispatcher a blocker for me in order to merge this. I'd love if we could rewrite it into something less global and have a small test for it.

I've removed the global singleton of connection provider and added tests for it.
Also did some refactor to address other comments.

**Comment by roobre** [comment]
LGTM, thanks for taking care of this!

**Pull request merged**